File: Platform/Workspace/Typescript/libs/workspace/memory/src/Population/Population.ts
import { RoleType, ObjectType, AssociationType, MetaPopulation } from '@allors/meta/system';

import { ChangeSet } from './ChangeSet';

export class Population {
  private readonly roleByAssociationByRoleType: Map<RoleType, Map<string, any>>;
  private readonly associationByRoleByAssociationType: Map<AssociationType, Map<string, any>>;

  private changedRoleByAssociationByRoleType: Map<RoleType, Map<string, any>>;
  private changedAssociationByRoleByAssociationType: Map<AssociationType, Map<string, any>>;

  private objects: string[];

  constructor(private readonly meta: MetaPopulation) {
    this.roleByAssociationByRoleType = new Map();
    this.associationByRoleByAssociationType = new Map();

    this.changedRoleByAssociationByRoleType = new Map();
    this.changedAssociationByRoleByAssociationType = new Map();
  }

  Snapshot(): ChangeSet {
    this.changedRoleByAssociationByRoleType.forEach((changedRoleByAssociation, roleType) => {
      const roleByAssociation = this.roleByAssociation(roleType);

      changedRoleByAssociation.forEach((changedRole, association) => {
        const originalRole = roleByAssociation.get(association);
        if (this.areSame(roleType.isOne, originalRole, changedRole)) {
          changedRoleByAssociation.delete(association);
        } else {
          roleByAssociation.set(association, changedRole);
        }
      });

      if (roleByAssociation.size == 0) {
        this.changedRoleByAssociationByRoleType.delete(roleType);
      }
    });

    this.changedAssociationByRoleByAssociationType.forEach((changedAssociationByRole, associationType) => {
      const associationByRole = this.associationByRole(associationType);
      changedAssociationByRole.forEach((changedAssociation, role) => {
        const originalAssociation = associationByRole.get(role);
        if (this.areSame(associationType.isOne, originalAssociation, changedAssociation)) {
          changedAssociationByRole.delete(role);
        } else {
          associationByRole.set(role, changedAssociation);
        }
      });
    });

    const snapshot = new ChangeSet(this.meta, this.changedRoleByAssociationByRoleType, this.changedAssociationByRoleByAssociationType);

    this.changedRoleByAssociationByRoleType = new Map();
    this.changedAssociationByRoleByAssociationType = new Map();

    return snapshot;
  }

  addObject(newObject: string): void {
    this.objects.push(newObject);
  }

  getRole(association: string, roleType: RoleType): any {
    const changedRole = this.changedRoleByAssociationByRoleType.get(roleType)?.get(association);
    if (changedRole !== undefined) {
      return changedRole;
    }

    return this.roleByAssociation(roleType).get(association);
  }

  // setRole(association: string,  roleType: RoleType, role: any): void
  // {
  //   // TODO: var normalizedRole = roleType.Normalize(role);
  //   const normalizedRole = role;
    
  //     if (roleType.isUnit)
  //     {
  //         // Role
  //         this.changedRoleByAssociation(roleType).set(association, normalizedRole);
  //     }
  //     else
  //     {
  //         var associationType = roleType.associationType;
  //         this.GetRole(association, roleType, out object previousRole);
  //         if (roleType.IsOne)
  //         {
  //             var roleObject = (DynamicObject)normalizedRole;
  //             this.GetAssociation(roleObject, associationType, out var previousAssociation);

  //             // Role
  //             var changedRoleByAssociation = this.ChangedRoleByAssociation(roleType);
  //             changedRoleByAssociation[association] = roleObject;

  //             // Association
  //             var changedAssociationByRole = this.ChangedAssociationByRole(associationType);
  //             if (associationType.IsOne)
  //             {
  //                 // One to One
  //                 var previousAssociationObject = (DynamicObject)previousAssociation;
  //                 if (previousAssociationObject != null)
  //                 {
  //                     changedRoleByAssociation[previousAssociationObject] = null;
  //                 }

  //                 if (previousRole != null)
  //                 {
  //                     var previousRoleObject = (DynamicObject)previousRole;
  //                     changedAssociationByRole[previousRoleObject] = null;
  //                 }

  //                 changedAssociationByRole[roleObject] = association;
  //             }
  //             else
  //             {
  //                 changedAssociationByRole[roleObject] = NullableArraySet.Remove(previousAssociation, roleObject);
  //             }
  //         }
  //         else
  //         {
  //             var roles = ((IEnumerable<DynamicObject>)normalizedRole)?.ToArray() ?? Array.Empty<DynamicObject>();
  //             var previousRoles = (DynamicObject[])previousRole ?? Array.Empty<DynamicObject>();

  //             // Use Diff (Add/Remove)
  //             var addedRoles = roles.Except(previousRoles);
  //             var removedRoles = previousRoles.Except(roles);

  //             foreach (var addedRole in addedRoles)
  //             {
  //                 this.AddRole(association, roleType, addedRole);
  //             }

  //             foreach (var removeRole in removedRoles)
  //             {
  //                 this.RemoveRole(association, roleType, removeRole);
  //             }
  //         }
  //     }
  // }

  private areSame(isOne: boolean, a: any, b: any): boolean {
    if (isOne) {
      return a === b;
    } else {
      if (a == null && b == null) {
        return true;
      } else {
        if (a == null || b == null) {
          return false;
        } else {
          return a.size === b.size && [...a].every((value) => b.has(value));
        }
      }
    }
  }

  private associationByRole(asscociationType: AssociationType): Map<string, object> {
    let associationByRole = this.associationByRoleByAssociationType.get(asscociationType);

    if (associationByRole === undefined) {
      associationByRole = new Map();
      this.associationByRoleByAssociationType.set(asscociationType, associationByRole);
    }

    return associationByRole;
  }

  private roleByAssociation(roleType: RoleType): Map<any, object> {
    let roleByAssociation = this.roleByAssociationByRoleType.get(roleType);

    if (roleByAssociation === undefined) {
      roleByAssociation = new Map();
      this.roleByAssociationByRoleType.set(roleType, roleByAssociation);
    }

    return roleByAssociation;
  }

  private changedAssociationByRole(associationType: AssociationType): Map<string, any> {
    let changedAssociationByRole = this.changedAssociationByRoleByAssociationType.get(associationType);

    if (changedAssociationByRole === undefined) {
      changedAssociationByRole = new Map();
      this.changedAssociationByRoleByAssociationType.set(associationType, changedAssociationByRole);
    }

    return changedAssociationByRole;
  }

  private changedRoleByAssociation(roleType: RoleType): Map<string, any> {
    let changedRoleByAssociation = this.changedRoleByAssociationByRoleType.get(roleType);

    if (changedRoleByAssociation === undefined) {
      changedRoleByAssociation = new Map();
      this.changedRoleByAssociationByRoleType.set(roleType, changedRoleByAssociation);
    }

    return changedRoleByAssociation;
  }
}

File: Platform/Workspace/CSharp/Adapters/Allors.Workspace.Adapters.Remote/Population/Population.cs
// <copyright file="DatabaseObject.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Remote
{
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using System.Linq;
    using Allors.Workspace.Meta;

    internal class Population
    {
        private readonly Dictionary<IRoleType, Dictionary<long, object>> roleByAssociationByRoleType;
        private readonly Dictionary<IAssociationType, Dictionary<long, object>> associationByRoleByAssociationType;

        private Dictionary<IRoleType, Dictionary<long, object>> changedRoleByAssociationByRoleType;
        private Dictionary<IAssociationType, Dictionary<long, object>> changedAssociationByRoleByAssociationType;

        internal Population()
        {
            this.roleByAssociationByRoleType = new Dictionary<IRoleType, Dictionary<long, object>>();
            this.associationByRoleByAssociationType = new Dictionary<IAssociationType, Dictionary<long, object>>();

            this.changedRoleByAssociationByRoleType = new Dictionary<IRoleType, Dictionary<long, object>>();
            this.changedAssociationByRoleByAssociationType = new Dictionary<IAssociationType, Dictionary<long, object>>();
        }

        internal ChangeSet Snapshot()
        {
            foreach (var roleType in this.changedRoleByAssociationByRoleType.Keys.ToArray())
            {
                var changedRoleByAssociation = this.changedRoleByAssociationByRoleType[roleType];
                var roleByAssociation = this.RoleByAssociation(roleType);

                foreach (var association in changedRoleByAssociation.Keys.ToArray())
                {
                    var role = changedRoleByAssociation[association];
                    roleByAssociation.TryGetValue(association, out var originalRole);

                    var areEqual = ReferenceEquals(originalRole, role) ||
                                   (roleType.IsOne && Equals(originalRole, role)) ||
                                   (roleType.IsMany && ((IStructuralEquatable)originalRole)?.Equals((IStructuralEquatable)role) == true);

                    if (areEqual)
                    {
                        changedRoleByAssociation.Remove(association);
                        continue;
                    }

                    roleByAssociation[association] = role;
                }

                if (roleByAssociation.Count == 0)
                {
                    this.changedRoleByAssociationByRoleType.Remove(roleType);
                }
            }

            foreach (var associationType in this.changedAssociationByRoleByAssociationType.Keys.ToArray())
            {
                var changedAssociationByRole = this.changedAssociationByRoleByAssociationType[associationType];
                var associationByRole = this.AssociationByRole(associationType);

                foreach (var role in changedAssociationByRole.Keys.ToArray())
                {
                    var changedAssociation = changedAssociationByRole[role];
                    associationByRole.TryGetValue(role, out var originalRole);

                    var areEqual = ReferenceEquals(originalRole, changedAssociation) ||
                                   (associationType.IsOne && Equals(originalRole, changedAssociation)) ||
                                   (associationType.IsMany && ((IStructuralEquatable)originalRole)?.Equals((IStructuralEquatable)changedAssociation) == true);

                    if (areEqual)
                    {
                        changedAssociationByRole.Remove(role);
                        continue;
                    }

                    associationByRole[role] = changedAssociation;
                }

                if (associationByRole.Count == 0)
                {
                    this.changedAssociationByRoleByAssociationType.Remove(associationType);
                }
            }

            var snapshot = new ChangeSet(this.changedRoleByAssociationByRoleType, this.changedAssociationByRoleByAssociationType);

            this.changedRoleByAssociationByRoleType = new Dictionary<IRoleType, Dictionary<long, object>>();
            this.changedAssociationByRoleByAssociationType = new Dictionary<IAssociationType, Dictionary<long, object>>();

            return snapshot;
        }

        internal void GetRole(long association, IRoleType roleType, out object role)
        {
            if (this.changedRoleByAssociationByRoleType.TryGetValue(roleType, out var changedRoleByAssociation) &&
                changedRoleByAssociation.TryGetValue(association, out role))
            {
                return;
            }

            this.RoleByAssociation(roleType).TryGetValue(association, out role);
        }

        internal void SetRole(long association, IRoleType roleType, object role)
        {
            if (role == null)
            {
                this.RemoveRole(association, roleType);
                return;
            }

            if (roleType.ObjectType.IsUnit)
            {
                // Role
                var unitRole = roleType.NormalizeUnit(role);
                this.ChangedRoleByAssociation(roleType)[association] = unitRole;
            }
            else
            {
                var associationType = roleType.AssociationType;
                this.GetRole(association, roleType, out var previousRole);
                if (roleType.IsOne)
                {
                    var compositeRole = roleType.NormalizeComposite(role);
                    this.GetAssociation(compositeRole, associationType, out var previousAssociation);

                    // Role
                    var changedRoleByAssociation = this.ChangedRoleByAssociation(roleType);
                    changedRoleByAssociation[association] = compositeRole;

                    // Association
                    var changedAssociationByRole = this.ChangedAssociationByRole(associationType);
                    if (associationType.IsOne)
                    {
                        // One to One
                        var previousAssociationObject = (long?)previousAssociation;
                        if (previousAssociationObject != null)
                        {
                            changedRoleByAssociation[previousAssociationObject.Value] = null;
                        }

                        if (previousRole != null)
                        {
                            var previousRoleObject = (long)previousRole;
                            changedAssociationByRole[previousRoleObject] = null;
                        }

                        changedAssociationByRole[compositeRole] = association;
                    }
                    else
                    {
                        changedAssociationByRole[compositeRole] = NullableArrayList.Remove(previousAssociation, compositeRole);
                    }
                }
                else
                {
                    var compositesRole = roleType.NormalizeComposites(role);
                    var previousRoles = (long[])previousRole ?? Array.Empty<long>();

                    // Use Diff (Add/Remove)
                    var addedRoles = compositesRole.Except(previousRoles);
                    var removedRoles = previousRoles.Except(compositesRole);

                    foreach (var addedRole in addedRoles)
                    {
                        this.AddRole(association, roleType, addedRole);
                    }

                    foreach (var removeRole in removedRoles)
                    {
                        this.RemoveRole(association, roleType, removeRole);
                    }
                }
            }
        }

        internal void AddRole(long association, IRoleType roleType, long role)
        {
            var associationType = roleType.AssociationType;
            this.GetAssociation(role, associationType, out var previousAssociation);

            // Role
            var changedRoleByAssociation = this.ChangedRoleByAssociation(roleType);
            this.GetRole(association, roleType, out var previousRole);
            var roleArray = (long?[])previousRole;
            roleArray = NullableArrayList.Add(roleArray, role);
            changedRoleByAssociation[association] = roleArray;

            // Association
            var changedAssociationByRole = this.ChangedAssociationByRole(associationType);
            if (associationType.IsOne)
            {
                // One to Many
                var previousAssociationObject = (long?)previousAssociation;
                if (previousAssociationObject != null)
                {
                    this.GetRole(previousAssociationObject.Value, roleType, out var previousAssociationRole);
                    changedRoleByAssociation[previousAssociationObject.Value] = NullableArrayList.Remove(previousAssociationRole, role);
                }

                changedAssociationByRole[role] = association;
            }
            else
            {
                // Many to Many
                changedAssociationByRole[role] = NullableArrayList.Add(previousAssociation, association);
            }
        }

        internal void RemoveRole(long association, IRoleType roleType, long role)
        {
            var associationType = roleType.AssociationType;
            this.GetAssociation(role, associationType, out var previousAssociation);

            this.GetRole(association, roleType, out var previousRole);
            if (previousRole != null)
            {
                // Role
                var changedRoleByAssociation = this.ChangedRoleByAssociation(roleType);
                changedRoleByAssociation[association] = NullableArrayList.Remove(previousRole, role);

                // Association
                var changedAssociationByRole = this.ChangedAssociationByRole(associationType);
                if (associationType.IsOne)
                {
                    // One to Many
                    changedAssociationByRole[role] = null;
                }
                else
                {
                    // Many to Many
                    changedAssociationByRole[role] = NullableArrayList.Add(previousAssociation, association);
                }
            }
        }

        internal void RemoveRole(long association, IRoleType roleType)
        {
            if (roleType.ObjectType.IsUnit)
            {
                // Role
                this.ChangedRoleByAssociation(roleType)[association] = null;
            }
            else
            {
                var associationType = roleType.AssociationType;
                this.GetRole(association, roleType, out var previousRole);

                if (roleType.IsOne)
                {
                    // Role
                    var changedRoleByAssociation = this.ChangedRoleByAssociation(roleType);
                    changedRoleByAssociation[association] = null;

                    // Association
                    var changedAssociationByRole = this.ChangedAssociationByRole(associationType);
                    if (associationType.IsOne)
                    {
                        // One to One
                        if (previousRole != null)
                        {
                            var previousRoleObject = (long)previousRole;
                            changedAssociationByRole[previousRoleObject] = null;
                        }
                    }
                }
                else
                {
                    var previousRoles = (long[])previousRole ?? Array.Empty<long>();

                    // Use Diff (Remove)
                    foreach (var removeRole in previousRoles)
                    {
                        this.RemoveRole(association, roleType, removeRole);
                    }
                }
            }
        }

        internal void GetAssociation(long role, IAssociationType associationType, out object association)
        {
            if (this.changedAssociationByRoleByAssociationType.TryGetValue(associationType, out var changedAssociationByRole) &&
                changedAssociationByRole.TryGetValue(role, out association))
            {
                return;
            }

            this.AssociationByRole(associationType).TryGetValue(role, out association);
        }

        private Dictionary<long, object> AssociationByRole(IAssociationType asscociationType)
        {
            if (!this.associationByRoleByAssociationType.TryGetValue(asscociationType, out var associationByRole))
            {
                associationByRole = new Dictionary<long, object>();
                this.associationByRoleByAssociationType[asscociationType] = associationByRole;
            }

            return associationByRole;
        }

        private Dictionary<long, object> RoleByAssociation(IRoleType roleType)
        {
            if (!this.roleByAssociationByRoleType.TryGetValue(roleType, out var roleByAssociation))
            {
                roleByAssociation = new Dictionary<long, object>();
                this.roleByAssociationByRoleType[roleType] = roleByAssociation;
            }

            return roleByAssociation;
        }

        private Dictionary<long, object> ChangedAssociationByRole(IAssociationType associationType)
        {
            if (!this.changedAssociationByRoleByAssociationType.TryGetValue(associationType, out var changedAssociationByRole))
            {
                changedAssociationByRole = new Dictionary<long, object>();
                this.changedAssociationByRoleByAssociationType[associationType] = changedAssociationByRole;
            }

            return changedAssociationByRole;
        }

        private Dictionary<long, object> ChangedRoleByAssociation(IRoleType roleType)
        {
            if (!this.changedRoleByAssociationByRoleType.TryGetValue(roleType, out var changedRoleByAssociation))
            {
                changedRoleByAssociation = new Dictionary<long, object>();
                this.changedRoleByAssociationByRoleType[roleType] = changedRoleByAssociation;
            }

            return changedRoleByAssociation;
        }
    }
}
